blockchain/vm/evm.go

```diff
@@ -538,6 +538,35 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
 	return evm.create(caller, codeAndHash, gas, value, contractAddr, humanReadable, codeFormat)
 }
 
+func (evm *EVM) GetPrecompiledContractMap(caller common.Address) map[common.Address]PrecompiledContract {
+	if codeInfo, ok := evm.StateDB.GetCodeInfo(caller); ok {
```

Comment:
I think we need to consider chain rule first.
And then if it is istanbul chain rule, the vm version should be considered.

Therefore how about this code?
```suggestion
	switch {
	case evm.chainRules.IsIstanbul:
		codeInfo := evm.StateDB.GetCodeInfo(caller)
		if codeInfo ==  VmVersionIstanbul {
			return PrecompiledContractsIstanbul
		}
		return PrecompiledContractsConstantinople
	default:
		return PrecompiledContractsConstantinople	
```

Reply:
First of all, I put the `GetCodeInfo` line at first because this function is not only used for smart contract addresses, but also with zero address, etc.
Istanbul chain rule info is used when allocating vmVersion info to the `currentBlockVMVersion` variable. BTW, I tried to extract the vmVersion logic from the chain rules switch because vmVersion there may be a difference of 2 or more.

Reply:
@KimKyungup I refactor the code as you suggested and also changed the naming of VmVersion. 
change naming of VmVersion https://github.com/klaytn/klaytn/pull/912/commits/afa456dfc109d315a894d5e29840a31f43598f3e
refactor GetPrecompiledContractMap https://github.com/klaytn/klaytn/pull/912/commits/db60f58712f54075d16abb4cfea97a8ef893bf0d